### ISSUE TITLE
project.getDirectories() and project.getDirectory('foo') don't finds directories when ancestors do not contain direct descendant source file

### ISSUE THREAD

**Issue by mitchlloyd** [dsherret/ts-simple-ast#285]
From reading the documentation I was expecting `project.getDirectories()` to return all of the directories in my project setup like this:

```
/dir1
  /dir2
    /dir3
      index.ts
```

When I run `project.getDirectories().map(d => d.getPath()` I would expect 3 entries:

```
/dir1
/dir1/dir2
/dir1/dir2/dir3
```

But instead I get 1:

```
/dir1/dir2/dir3
```

`project.getDirectory('foo')` cannot find dir2 either. This throws:

```
project.getDirectoryOrThrow('dir1/dir2');
```

And this succeeds:

```
project.getDirectoryOrThow('dir1/dir2/dir3');
```

However if I add a source file to dir1 then I get the results I expect.

```
/dir1
  index.ts
  /dir2
    /dir3
      index.ts
```

This only comes up if you have a directory structure like this:

```
/src
  /components
    /foo-component
    /bar-component
  /utils
    /foo-util
    /bar-util
```

After digging into this I'm realizing that in practice this gotcha would be extremely rare since it's common to have an index.ts file in your `/src` directory as an entry point.  For me this was a surprise as I was setting up a test project to try out ts-simple-ast.

**Comment by dsherret** [comment]
Hey @mitchlloyd, in version 10 that's somewhat fixed... `project.getDirectory('dir1/dir2');` and `project.getDirectory('dir1')` will not throw (see #268).

Yes, the current design is to not add any directories that don't have source files. That might be a bad design though.

How are you setting up the project object? Are you using a tsconfig.json or something else? (Btw, based on this, I just opened #286)

**Comment by mitchlloyd** [comment]
Just getting start and so far my code looks like this 

```ts
import Project from 'ts-simple-ast';

const project = new Project({
  addFilesFromTsConfig: false,
});

project.addExistingSourceFiles('fixture/src/**/*.{ts,tsx}');

const pageDirectory = project.getDirectoryOrThrow('fixture/src/pages');
```

I'm spiking out a script to modify some files and I created a "fixture" directory in my project to test things out:

```
fixture/
├── src
│   ├── pages
│   │   ├── home
│   │   │   └── components
│   │   │       └── home
│   │   │           ├── component.tsx
│   │   │           └── index.ts
│   │   └── ignored
│   │       └── components
│   │           └── ignored
│   │               ├── component.tsx
│   │               └── index.ts
└── tsconfig.json
```

**Comment by mitchlloyd** [comment]
Updated my example code in the original post. I meant to use `getDirectoryOrThrow` to demonstrate that the middle directory could not be found but instead I had `getDirectory`. I'm using version 9.5.0.

**Comment by dsherret** [comment]
Ok, I just opened #287. Will that solve your case?

By the way, until these issues are implemented, if you only need to modify source files then you could forget with dealing with directories and instead just get the source files in that pages directory:

```ts
project.addExistingSourceFiles('fixture/src/**/*.{ts,tsx}');

const pagesSourceFiles = project.getSourceFiles('fixture/src/pages/**/*.{ts,tsx}');
```

Or alternatively, just get the source files returned from: 

```ts
const addedSourceFiles = project.addExistingSourceFiles('fixture/src/**/*.{ts,tsx}');
```

**Comment by dsherret** [comment]
The two issues are implemented for v10. I'll release it sometime this week (probably tomorrow) and will comment back here then.

**Comment by dsherret** [comment]
This should be fixed in 10.0.1. Let me know if the behaviour needs some tweaking or if there are any issues.

Also, thanks for reporting this. I think it's much better now.

**Comment by mitchlloyd** [comment]
I just checked this change in version 10.0.3 against my project. All the directories that I expected to see originally are now there! Hopefully this will save someone some confusion in the future.
